### PR TITLE
Add StakeLockContract with basic locking/unlocking based on epoch

### PR DESCRIPTION
**Pull request by rlan35**
We need a staking contract that support more functionalities such as locking based on epochs.  Will have another PR to add batch query func.

### CODE REVIEW

File: contracts/StakeLockContract.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * @dev Error messages for require statements
+     */
+    string internal constant ALREADY_LOCKED = 'Tokens already locked';
+    string internal constant NOT_LOCKED = 'No tokens locked';
+    string internal constant AMOUNT_ZERO = 'Amount can not be 0';
+
+    uint256 internal constant LOCK_PERIOD_IN_EPOCHS = 3;
+
+    uint256 internal numBlocksPerEpoch = 5;
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: this is for testing only, please add a comment.

File: contracts/StakeLockContract.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+contract StakeLockContract {
+    /**
+     * @dev Error messages for require statements
+     */
+    string internal constant ALREADY_LOCKED = 'Tokens already locked';
+    string internal constant NOT_LOCKED = 'No tokens locked';
+    string internal constant AMOUNT_ZERO = 'Amount can not be 0';
+
+    uint256 internal constant LOCK_PERIOD_IN_EPOCHS = 3;
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: why 3 here? any comment?

File: contracts/StakeLockContract.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+     *      a specified address
+     */
+    mapping(address => lockedToken) private locked;
+    mapping(address => uint) private indices;
+    address[] private accountList;
+
+    event Locked(address indexed _of, uint _amount, uint256 _epoch);
+    event Unlocked(address indexed account, uint index);
+
+	/**
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: nit: alignment

File: contracts/StakeLockContract.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    function lock(uint256 _epoch)
+        public
+        payable
+        returns (bool)
+    {
+        // If tokens are already locked, then functions extendLock or
+        // increaseLockAmount should be used to make any changes
+        require(lockedBalanceOf(msg.sender) == 0, ALREADY_LOCKED);
+        require(msg.value != 0, AMOUNT_ZERO);
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: != or > ?

File: contracts/StakeLockContract.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     * @dev Unlocks the unlockable tokens of a specified address
+     * @param _requestor Address of user, claiming back unlockable tokens
+     */
+    function unlock(address payable _requestor)
+        public
+        returns (uint256 unlockableTokens)
+    {
+        unlockableTokens = getUnlockableTokens(_requestor);
+
+        if (unlockableTokens != 0) {
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: != 0 or >= 0?

File: contracts/StakeLockContract.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * @dev Gets the epoch number of the specified block number.
+     */
+    function currentEpoch()
+        public
+        view
+        returns (uint256)
+    {
+        return block.number / numBlocksPerEpoch;
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: any risk of overflow?

File: contracts/StakeLockContract.sol [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        // If tokens are already locked, then functions extendLock or
+        // increaseLockAmount should be used to make any changes
+        require(lockedBalanceOf(msg.sender) == 0, ALREADY_LOCKED);
+        require(msg.value != 0, AMOUNT_ZERO);
+
+        locked[msg.sender] = lockedToken(msg.value, block.number, currentEpoch(), 1);
+
+        emit Locked(msg.sender, msg.value, _epoch);
+        return true;
```
Comment (LeoHChen) [inline contracts/StakeLockContract.sol]: in which condition this function will return false?

File: contracts/StakeLockContract.sol (hunk 2 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: Other projects like Eth also lock for something like 3 epoch. It's just a good number to lock. Not too short not too long. We will decide the final number later (added comment)

File: contracts/StakeLockContract.sol (hunk 1 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: Done.

File: contracts/StakeLockContract.sol (hunk 3 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: Done.

File: contracts/StakeLockContract.sol (hunk 4 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: Changed to > to be safer.

File: contracts/StakeLockContract.sol (hunk 5 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: the value is uint

File: contracts/StakeLockContract.sol (hunk 6 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: the two number are all uint256. As long as numBlocksPerEpoch is not 0, it should be fine.

File: contracts/StakeLockContract.sol (hunk 7 above)
Comment (rlan35) [reply contracts/StakeLockContract.sol]: This is a previous common practice to return true when execution succeeded, tho there is no false case.

**Comment by LeoHChen** [comment]
Please add some testing log to the PR. Thanks.

File: contracts/StakeLockContract.sol [hunk 8]
```diff
@@ -8,9 +8,9 @@ contract StakeLockContract {
     string internal constant NOT_LOCKED = 'No tokens locked';
     string internal constant AMOUNT_ZERO = 'Amount can not be 0';
 
-    uint256 internal constant LOCK_PERIOD_IN_EPOCHS = 3;
+    uint256 internal constant LOCK_PERIOD_IN_EPOCHS = 3;  // Final locking period TBD.
 
-    uint256 internal numBlocksPerEpoch = 5;
+    uint256 internal numBlocksPerEpoch = 5;  // This value is for testing only
```
Comment (nicolasburtey) [inline contracts/StakeLockContract.sol]: maybe use FIXME keyword to be able to parse it later easily?

**Pull request merged**